### PR TITLE
Refuse h2 streams that can't be scheduled

### PR DESCRIPTION
**Pull request by Dridi**
The initial plan was a one-liner to get rid of an `XXXAZ` assert (2nd patch) but testing this case proved to be hard because of a bug in the thread pools (fixed in 1st patch) and while I was playing a bit with h2 resets I realized that the `sess_drop` counters includes streams so I introduced a new `req_drop` counter (3rd patch).

I expect the 3rd patch to be potentially controversial, but I tried to keep the changes to a minimum.

As a side note, since we are breaking libvarnishapi, it may be a good occasion to sort the counters. For example `sess_drop` and `sess_dropped` could be next to each other.

### CODE REVIEW

**Comment by Dridi** [comment]
Fixes #2311

**Comment by mbgrydeland** [comment]
For the counter patch, we have two option as I see it:

1. Docfix the sess_drop / rename it, to be the sum of scheduling errors (which could be H1 sessions, and H2 sessions or streams). Then we don't need to special case the scheduling type, and just increase ndropped also for these scheduling errors.

2. Keep the H1 and H2 errors counted separately. I am not sure I see the value of keeping these errors separate though, as they both draw from the same resource pool (the threads), and increasing the pool size is the possible action in both cases. If we go this way, the separate counter should be included in the pools locking scheme like ndropped is today, instead of adding a new lock type.

**Comment by mbgrydeland** [comment]
I agree that the REFUSED_STREAM is the right response to send to the client on these types of errors. It is correct to have done the HPACK processing even though we are not replying on the stream, to keep up with the clients HPACK state.

**Comment by Dridi** [comment]
> For the counter patch, we have two option as I see it:

I went for solution 2 because `sess_drop` isn't accurate for h2 streams. Depending on the consensus I will update the 3rd patch accordingly. Now that you mention it, the separate counter for `req_drop` is so obvious that it's embarrassing :)

**Comment by Dridi** [comment]
I just realized that `sess_drop` isn't the correct homologous counter, so I just pushed a patch going for solution 2 again, taking @mbgrydeland's remarks into account, and introducing `req_dropped` instead.

**Comment by Dridi** [comment]
I just realized that `sess_drop` isn't the correct homologous counter, so I just pushed a patch going for solution 2 again, taking @mbgrydeland's remarks into account, and introducing `req_dropped` instead.

File: bin/varnishd/main.vsc [hunk 1]
```diff
@@ -240,8 +240,14 @@
...
-	Number of times session was dropped because the queue were too long
-	already. See also parameter thread_queue_limit.
+	Number of times an HTTP/1 session was dropped because the queues were
+	too long already. See also parameter thread_queue_limit.
+
+.. varnish_vsc:: req_dropped
+	:oneliner:	Requests dropped
+
+	Number of times an HTTP/2 stream was refused because the queues were
+	too long already. See also parameter thread_queue_limit.
```
Comment (fgsch) [inline bin/varnishd/main.vsc]: This sentence sounds weird. The "queue was full" or something similar would be better.

**Comment by Dridi** [comment]
@fgsch thanks, now updated. It wasn't a good idea to make this change that late :)

**Comment by bsdphk** [comment]
Tested on FreeBSD, looks OK to me.

**Comment by daghf** [comment]
@Dridi, would you mind also having a look at the `XXXAZ(Pool_Task(..))` in `h2_ou_session` ?

**Pull request merged**